Update docs with Provider configuration (#1815)
nasa/cumulus: docs/operator-docs/provider.md

---
id: provider
title: Provider Configuration
hide_title: true
---

# Provider Configuration

> In Cumulus a Provider represents a endpoint from which data is ingested.   For example, a http server serving data from a data provider, or an S3 bucket deployed within your organization with data staged for ingest into Cumulus.

Reviewer: ```suggestion
In Cumulus, a Provider represents a endpoint from which data is ingested.   For example, a HTTP server serving data from a data provider, or an S3 bucket deployed within your organization with data staged for ingest into Cumulus.
```

Please note:

> * While *connection* configuration is defined here, things that are more specific to a specific ingest setup (e.g. 'What target directory should we be pulling from' or 'How is duplicate handling configured?') are generally defined in a Rule or Collection, not the Provider.

Reviewer: Going back and forth but I think it would be nice if "rule" and "collection" linked to the page describing them

Author: I think I agree, but those pages don't exist in the Operator docs yet (and writing them feels ... out of scope?) , and otherwise the explicit description only exists in the Cookbooks Intro page.... which we're linking to from the Glossary. 

But the entire point of this ticket is that the Cookbook oriented example isn't a comprehensive configuration document, and we appear to need that given user/operator feedback on that point. 

What do you think if I write the ticket for Rules/Collections like this one and include cross linking in the AC?    Else write a stub that refers back to the Cookbook intro?   

🤔

Reviewer: I guess I was just thinking of linking to the cookbook intro page for now. But given the additional complexity I think we can skip it for now.

I really think the definition of rules/providers/collections needs to be done much more uip front in our docs, like a "Getting started" section, but that's definitely out of scope

> * Task-specific provider configurations that are expected to be configured on a *per-workflow* basis exist, and may be tweaked to allow for task-specific configuration (e.g. https timeout values for the `discover-granules` task).

Reviewer: I'm not really sure what this is trying to say

Author: Ouch.  

This is trying to end-run Operators who come in looking to tweak connection behaviors that aren't intended for the Provider scope.    For example, we have several things like http/https timeouts that are *task* configurations that you might expect to be Provider/global oriented. 

How about 
```suggestion
* Task-specific provider configurations that are expected to be configured on a *per-workflow* basis exist, these are generally configured as part of the workflow definition and are documented in task documentation/README).
```

Reviewer: ```suggestion
* There is some provider behavior which is controlled by task-specific configuration and not the provider definition. This configuration has to be set on a **per-workflow** basis. For example, see the [`httpListTimeout` configuration on the `discover-granules` task](https://github.com/nasa/cumulus/blob/master/tasks/discover-granules/schemas/config.json#L84)
```

Author: 👍  I can live with that.

A Provider can be created via use of the [API](https://nasa.github.io/cumulus-api/#create-provider) or via a client application like the [Cumulus Dashboard](https://github.com/nasa/cumulus-dashboard).

## Provider Configuration

> The Provider configuration is defined by a JSON object that takes different configuration keys depending on the provider type.    The following are definitions of a typical configuration values relevant for the various providers:

Reviewer: ```suggestion
The Provider configuration is defined by a JSON object that takes different configuration keys depending on the provider type.    The following are definitions of typical configuration values relevant for the various providers:
```

### S3

|Key  |Type |Required|Description|
|:---:|:----|:------:|-----------|
|id|string|Yes|Unique identifier for the provider|

> |globalConnectionLimit|integer|No|Integer specifying the connection limit for the provider. This is the maximum number of connections Cumulus compatible ingest lambdas are expected to make to a provider.  Defaults to unlimited |

Reviewer: Is this relevant from S3?

Author: I believe so, yes.     There's nothing in `ingest/src/lock.proceed` that cares about the provider type, nor in sync-granule (which is the only task to use that setup). 

That raises the point that `Cumulus compatible ingest lambdas` is a bit.... mushy as currently that's only SyncGranules.      DiscoverGranules/PDR doesn't care because it's periodicity is controlled by the Rule setting and/or the lambda timeout.     🤔     Perhaps we should write a ticket or at least discuss if discovery should respect this limit as it's not terribly global at the moment.

Reviewer: As long as it is relevant for S3, the other concerns are secondary for now

> |protocol|string|Yes|The protocol for this provider. Must be `S3` for this provider type. |

Reviewer: I think the ENUM in the schema is case sensitive

```suggestion
|protocol|string|Yes|The protocol for this provider. Must be `s3` for this provider type. |
```
